### PR TITLE
Walk, Transform, Diff, Equal, and String.

### PR DESCRIPTION
**Pull request by paddycarver**
This turned into a massive commit. I'm really sorry.

I set out to implement Walk and Transform. Walk was simple enough.
Transform was implemented quickly and easily enough. And then I tried to
test it.

Testing Transform reasonably necessitated all the other changes you see
here:

* We needed to be able to Diff Values to be able to express changes
  between Values so we could test that Transform made the changes we
  wanted it to.
* We needed to be able to represent AttributePaths as strings so we
  could have specific, useful test output. This also entailed making
  Values strings (AttributePaths can use Values as a step) and making
  Types have more specific strings (aggregate types now expose the
  subtypes that make up their type signature).
* In doing so, we kept accidentally overwriting our AttributePaths,
  meaning it was time for them to stop being modified in place and start
  being passed by value, with the steps defensively copied to avoid
  accidental modification.
* We needed the ability to detect whether Diffs were equal to know
  whether the changes we expected were the changes we actually got or
  not.
* We needed to make Values an AttributePathStepper so we could implement
  Diff, which is done by Walking one Value, and then retrieving the same
  sub-Value from the Value we're Diffing against to check if they're the
  same. We already have WalkAttributePath to retrieve part of a Go type
  based on the AttributePath passed in, we just needed to tell it how to
  index into a Value type.
* We needed the ability to Copy a Value type so that it was not sharing
  the same memory space anymore so our transformations didn't leak
  across tests.

The result is this PR of alarming size. The good news is a lot of these
new lines are tests.

### CODE REVIEW

**Comment by paddycarver** [comment]
A lot of this code needs GoDoc strings. More tests are also needed, as are changelog entries.

These things are all forthcoming before we merge this, I just wanted to get the working code up so @alexsomesan could see and play with it.

**Comment by alexsomesan** [comment]
Wow! Is it Christmas all over again?! This PR is just full of juicy goodies! 
I especially appreciate the changes to AttributePath to return values and not mutate itself. I really wish there .Errorf would expose the path somehow too. Although I guess I could just use it's String() method now to get to it.

Regardless, this is awesome. Thanks!
I will test these tomorrow and report impressions :)

**Comment by paddycarver** [comment]
I am anticipating a separate PR once this is merged to make AttributePathError a bit more useful. Just trying to cut scope on this PR when I can 😂

**Comment by alexsomesan** [comment]
Changes in #58 will require updates to this PR if 58 merges first or the other way around, because walk and diff both check types via Value.Is(...) which was removed in #58.

**Comment by paddycarver** [comment]
Yup! Scattered some TODOs around this to take 58 into account. :)

**Comment by paddycarver** [comment]
Fixes #53.

**Comment by paddycarver** [comment]
Rebased after #58 and updated to use its Type() goodies.

File: tfprotov5/tftypes/attribute_path.go
Comment (paultyng) [inline tfprotov5/tftypes/attribute_path.go:84]: I think you could probably just collapse these down in to something like the below? The casting seems unnecessary and already handled by the equals? If value's aren't comparable though makes sense to pull that case out.

```go
default:
  if aStep != oStep {
    return false
  }
```

**Review by paultyng — APPROVED** [review]
Minor comment, but LGTM (once testing coverage added)

**Review by kmoe — APPROVED** [review]

File: tfprotov5/tftypes/walk.go
Comment (alexsomesan) [inline tfprotov5/tftypes/walk.go:54]: Thought experiment: I wonder if this should somehow surface that the visitation was interrupted before a complete exploration of all nodes could be finished. A special type of error maybe?

**Review by alexsomesan — APPROVED** [review]
Looks great!

I've actually made use of  Transform in the Kubernetes alpha provider and it works as expected. 

I did try to use Diff in a few places but it errors out with `incompatible types` when encountering dissimilar tuples. Because TF core always parses lists as tuples, it means any two lists with distinct amount of elements will make Diff err out. I don't think that's an issue with Diff itself, but rather an unfortunate combination of factors in my use-case.

File: tfprotov5/tftypes/walk.go
Comment (paddycarver) [reply tfprotov5/tftypes/walk.go:54]: This seems to mirror the Go standard library walk behaviors, in the expectation that the original caller has some control over the callback? And given the callback is the one returning `shouldContinue`, they should know some things may be skipped? Similarly, the callback can return its own error that the caller can handle. I guess I'm not clear on when this would be used, or why.

**Comment by paddycarver** [comment]
OK, finally finished adding all the tests on this. I think we're ready for final review and merge. Apologies for the monster diff. :( This kinda spiraled out of control.

**Comment by paddycarver** [comment]
Fixes #53.
Fixes #56.

**Review by kmoe — APPROVED** [review]
Excellent! LGTM

**Pull request merged**
